### DISCUSSION TITLE
How to use DoUpdate with a TypeConverter field

### DISCUSSION THREAD

**Discussion by JeremyLWright** [simolus3/drift#1556] [Q&A] [answered]
I'm trying to "upsert" using the onConflict feature. However, I'm applying the upsert against a table with a TypeConverter on the column. This question may be related to #1158. When I try to use the Companion.custom() object to resolve the conflict, I get the following error message

> Allowed parameters must either be null or an int, num, String or List<int>.: Instance of 'Int64'

Here is my minimal example

```
import 'package:drift/drift.dart';
import 'package:fixnum/fixnum.dart';

part 'minimal_example.g.dart';

class Int64Converter extends TypeConverter<Int64, String> {
  const Int64Converter();

  @override
  Int64? mapToDart(String? fromDb) {
    if (fromDb != null) {
      return Int64.parseInt(fromDb);
    }
    return null;
  }

  @override
  String? mapToSql(Int64? value) {
    return value?.toString().padLeft(19, '0');
  }
}

class NumberTable extends Table {
  TextColumn get number => text().map(const Int64Converter())();
  @override
  Set<Column> get primaryKey => {number};
}

@DriftDatabase(tables: [NumberTable], daos: [NumberDao])
class Database extends _$Database {
  Database(QueryExecutor e) : super(e);

  @override
  int get schemaVersion => 1;
}

@DriftAccessor(tables: [NumberTable])
class NumberDao extends DatabaseAccessor<Database> with _$NumberDaoMixin {
  NumberDao(Database db) : super(db);

  Future<void> increment(Int64 n) {
    return into(numberTable).insert(NumberTableData(number: n), onConflict: DoUpdate((old) {
      return NumberTableCompanion.custom(number: Variable(Int64.TWO));
    }));
  }
}

```

Driver

```
void main(List<String> arguments) async {
  final db = Database(NativeDatabase.memory());
  final dao = NumberDao(db);
  print('Insert 1');
  await dao.increment(Int64.ONE);
  print('Insert 1 again');
  await dao.increment(Int64.ONE);
  print('Done');
}
```

It seems that the Companion objects don't leverage the TypeConverter, but I cannot find how to proceed with the update.

**Comment by simolus3** [comment, marked as answer]
It's true that the `.custom` companion constructor doesn't support type converters. It can't, really: The `Expression` passed can be any SQL value and it's not obvious how a type converter might be applied to that. Ideally, the companion should expect an `Expression<int>` here as that's the type in the database. This was an oversight from my side, but it's a breaking change to fix.

So, until the next major release, it looks like you'll have to work around this with `number: Variable(2).dartCast()`. Sorry for the inconvenience, it will get better :D

**Reply by JeremyLWright** [reply]
Thank you for the response.

This library is saving us massive time. We're incredibly thankful for such a high quality library. Thank you. Small oversights are natural. No need to apologize. 🥇 

However, I don't understand how to apply dartCast() here. My database has a string value in it. My typeconverter converts fixnum::Int64 to a zero-padded string ``class Int64Converter extends TypeConverter<Int64, String>``

<img width="1272" alt="Screen Shot 2021-11-23 at 12 51 17 PM" src="https://user-images.githubusercontent.com/2302957/143093978-cd367a4a-fcab-46cd-b2ef-1e6257f54e94.png">


.

**Reply by simolus3** [reply]
Thanks! And I should have read your code better :D You could use `Variable<String>('2').dartCast<Int64>()` in that case. The `dartCast` just exists so that we can use the right type for the database (which, in your case, it a `String`) and ignore the warnings from drift (which wrongly expects a `Int64`).

You could also apply your converter explicitly: `Variable(Int64Converter.mapToSql(Int64.TWO).dartCast())`.

**Reply by JeremyLWright** [reply]
That works! Thank you!!
